pkcs12/src/kdf.rs

```diff
@@ -0,0 +1,108 @@
+//! Implementation of the key derivation function
+//! [RFC 7292 Appendix B](https://datatracker.ietf.org/doc/html/rfc7292#appendix-B)
+
+use alloc::vec::Vec;
+use digest::{core_api::BlockSizeUser, Digest, FixedOutputReset, OutputSizeUser, Update};
+use zeroize::Zeroize;
+
+/// Transform a utf-8 string in a unicode (utf16) string as binary array.
+/// The Utf16 code points are stored in big endian format with two trailing zero bytes.
+fn str_to_unicode(utf8_str: &str) -> Vec<u8> {
```

Comment:
The RFC talks about `BMPString` which happens to be UTF16, but renaming that `str_to_bmpstring` or something would be great I think.

Reply:
Hmm, I guess we should really add `BMPString` support to `der`

Reply:
I'm a bit uncomfortable with the use of `str` for inputs here. This will backfire with any utf8mb4 character. https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c4edb02a6d7bf8fb45147475e43f1930

I don't know what the best course of action is here though, the best I can think of would be to add a `der::asn1::BmpString` but it is meant to only be subtyped from what I can read (https://www.oss.com/asn1/resources/asn1-made-simple/asn1-quick-reference/bmpstring.html)

Reply:
(sorry last comment was sent before I refreshed the page, well, go for a `BmpString` then)

Reply:
The format of an ASN1 `BMPString` differs slightly from what is here required: It has a leading id byte `0x1e`, one or bytes indicating the length in bytes, and no terminating zeros. I.e. using `BMPString` would also requires some pre-processing. Multi-Byte UTF8 characters should work fine here, at least the example you provided gives the exact same result als openssl's function `PKCS12_gen_key_utf8` (see the added test case and the respective test program).

Openssl provides similar functions for `PKCS12_gen_key_uni` and `PCKS12_gen_asc` for unicode and ASCII string passwords (see here: [https://www.openssl.org/docs/man3.0/man3/PKCS12_key_gen_utf8.html](https://www.openssl.org/docs/man3.0/man3/PKCS12_key_gen_utf8.html)) Maybe a similar approach could be taken here.

However, I am not sure how to interpret the remark "There are no Unicode byte order marks." in the RFC.

Reply:
[last commit](https://github.com/RustCrypto/formats/pull/1154/commits/282baa59bac74845c280308877f291a57559933f) does not really fix the issue with unicode characters that do not encode on two bytes.
[RFC 7292](https://datatracker.ietf.org/doc/html/rfc7292#appendix-B.1) does not really specify the behavior with those.
And I would be in favor of outright rejecting them (or maybe hide a `From<Vec<u8>>` under `hazmat` feature flag for a `der::asn1::BmpString`).

Reply:
A BOM would precede the string, and if it had one in this case would be `FEFF` to indicate big endian.

`BmpString` (if it existed) would still be a nice place to handle all of the string conversions. The header doesn't matter: we can still use [`EncodeValue`](https://docs.rs/der/latest/der/trait.EncodeValue.html) to avoid it, and two trailing digest bytes are easily added to the `Digest` prior to finalization.

We can worry about that all later though. It isn't needed for an initial PR.